lib/srv/alpnproxy/conn_test.go

```diff
@@ -0,0 +1,361 @@
+			require.NoError(t, err)
+			require.Equal(t, dataWritten, buf[:n])
+		}
+	})
+
+	t.Run("LargeContent", func(t *testing.T) {
+		t.Run("SingleRead", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			r, w := makePingConn(t)
+			defer r.Close()
+			defer w.Close()
+
+			largeSlice := make([]byte, math.MaxUint32)
+			errChan := make(chan error, 2)
+
+			// Produce some data to it to match after reading.
+			written, err := rand.Read(largeSlice)
+			require.NoError(t, err)
+			require.Equal(t, math.MaxUint32, written)
+
+			// Write goroutine.
+			go func() {
+				_, err = w.Write(largeSlice)
+				errChan <- err
+			}()
+
+			// Read goroutine.
+			go func() {
+				buf := make([]byte, math.MaxUint32)
+				read, err := r.Read(buf)
+				if err != nil {
+					errChan <- fmt.Errorf("expected no read error but got %s", err)
+					return
+				}
+
+				if read != math.MaxUint32 {
+					errChan <- fmt.Errorf("expected to read %d but got %d", math.MaxUint32, read)
+					return
+				}
+
+				if !bytes.Equal(largeSlice, buf) {
+					errChan <- fmt.Errorf("expected to content to be the same")
+					return
+				}
+			}()
+
+			for i := 0; i < 1; i++ {
+				select {
+				case <-ctx.Done():
+					require.Fail(t, "exceeded time to read/write message")
+				case err := <-errChan:
+					require.NoError(t, err)
+				}
+			}
+		})
+
+		t.Run("LargerThanUint32", func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			errChan := make(chan error)
+			_, w := makePingConn(t)
+			defer w.Close()
+
+			// Place into a goroutine to avoid freezing tests in case where the
+			// write succeeds.
+			go func() {
+				largeSlice := make([]byte, math.MaxUint32+1)
+				_, err := w.Write(largeSlice)
+				errChan <- err
+			}()
+
+			select {
+			case err := <-errChan:
+				require.True(t, trace.IsBadParameter(err), "expected error to be `BadParameter` but got %T", err)
+			case <-ctx.Done():
+				require.Fail(t, "expected to receive the write error but nothing was received")
+			}
+		})
+	})
+}
+
+// makePingConn creates a piped ping connection.
+func makePingConn(t *testing.T) (*pingConn, *pingConn) {
+	t.Helper()
+
+	writer, reader := net.Pipe()
+	return &pingConn{Conn: writer}, &pingConn{Conn: reader}
+}
+
+// makeTCPPingConn creates TCP connections to have asynchronous read/write.
+func makeTCPPingConn(t *testing.T) (*pingConn, *pingConn) {
+	t.Helper()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+	t.Cleanup(func() { listener.Close() })
+
+	connChan := make(chan net.Conn, 2)
+
+	// Accept routine.
+	go func() {
+		conn, err := listener.Accept()
+		if err != nil {
+			cancel()
+			return
+		}
+		connChan <- conn
+	}()
+
+	// Dial routine.
+	go func() {
+		conn, err := net.Dial("tcp", listener.Addr().String())
```

Comment:
is it possible to use something like `net.Pipe` instead of dial?

Reply:
Most of the tests are using `net.Pipe` connections. Only one test uses TCP connections to test the "asynchronous" Read/Write, where everything is written to the connection and later is consumed. I've tried to do this with the `bufio` package, but it leads to implementing a `net.Conn` with buffering inside. It felt a lot more code was written just to test this specific scenario, where TCP connections provide everything needed.

Reply:
cool. if it's a lot of effort then not worth it

just curious if this would work together with a buffer?
https://github.com/gravitational/teleport/blob/edc29051ec17aead90a6a76cabd43446472d2afc/lib/utils/pipenetconn.go#L26-L41

Reply:
I've tried this connection with `bufio` Reader/Writer, and it didn't work well. But thinking now, using just regular `bytes.Buffer` with it might do the job.

Reply:
It worked well. Thanks again!